[av] Fix precision of the player position

jpudysz opened the pull request:
# Why

Hello, Expo Team!

We are developing a complex video editor for our client, utilizing your `EXAVPlayer`. This editor enables users to seek any millisecond of the video. However, we encountered some issues while attempting to use `setPositionAsync`.

# How

As per Apple's documentation (https://developer.apple.com/documentation/avfoundation/avplayer/1387741-seektotime?language=objc), we understand that `toleranceBefore` and `toleranceAfter` can be set. Despite this, we've noticed edge cases where EXAVPlayer fails to navigate to the specified millisecond.

# Test Plan

The issue was tested in a local project. Adjusting the precision resolved the problem related to rounding.

Before:
1. Load any video into the `<Video />` tag, providing a reference (`videoRef`).
2. Execute the following:

```tsx
videoRef.current?.setPositionAsync(267, {
    toleranceMillisBefore: 0,
    toleranceMillisAfter: 0
})
```
3.  Observe `newPosition` in Objective-C.

![image](https://github.com/expo/expo/assets/9088288/37ce99d9-c564-467f-86ed-58cc8ef504bf)

4. Notice the video did not seek to 267 ms

After:

1. Load any video into the `<Video />` tag, providing a reference (`videoRef`).
2. Execute the following:

```tsx
videoRef.current?.setPositionAsync(267, {
    toleranceMillisBefore: 0,
    toleranceMillisAfter: 0
})
```
3.  Observe `newPosition` in Objective-C.

![image](https://github.com/expo/expo/assets/9088288/eb774e05-3fe1-4fa7-9702-ba36a9e74c1a)


4. The video now seeks accurately to 267 ms.

# Checklist

No need to update the docs

jpudysz committed:
fix: precission of newPosition

tsapeta reviewed (approved):
Good catch! I'll make sure to include it in SDK 50.
Congrats on your first contribution 👏 😉

jpudysz commented:
Thank you, @tsapeta! I'm looking forward to giving you a high five at the next App.js Conference 😁

tomekzaw reviewed (commented):
Good catch!

jpudysz committed:
Merge branch 'expo:main' into main

jpudysz committed:
feat: add option for zero tolerance

jpudysz commented:
Hey, I've added another case and updated the PR – please let me know your thoughts on it and if I should open new PR.

While the previous fix resolves my issue for a single video, I've noticed that `EXAVPlayer` sometimes disrupts the synchronization when handling two videos. This issue arises because one line of the code compares a JavaScript number to a `long` value. While rounding here might seem convenient, it occasionally leads to synchronization problems.

```objc
// compare number (int) to long
mustSeek = currentPositionMillis.longValue != [self _getRoundedMillisFromCMTime:_currentPosition].longValue;
```

The proposed solution is to always seek if the user sets the tolerance to zero. This implies that the user acknowledges the potential performance drawbacks. With this adjustment, I can consistently synchronize my two videos down to the exact millisecond.

Again tested locally CC @tsapeta

jpudysz committed:
Merge branch 'expo:main' into main

jpudysz committed:
chore: update manually changelog

jpudysz committed:
Merge branch 'expo:main' into main

jpudysz committed:
fix: conversion to NSNumber

jpudysz commented:
One more case guys! 😄 

# Why

With the two changes mentioned above, everything works perfectly, and `self.player.currentTime` is always set to the desired millisecond.

Unfortunately, sending the event back to JavaScript involves a different rounding algorithm. Even though the player successfully seeks to the desired millisecond, the reported time is slightly different.

Problem lives in the `_getRoundedMillisFromCMTime` where the value is converted from `CMTime` to `NSNumber`.

```objc
- (NSNumber *)_getRoundedMillisFromCMTime:(CMTime)time
{
  return CMTIME_IS_INVALID(time) || CMTIME_IS_INDEFINITE(time) ? nil : @((long)(CMTimeGetSeconds(time) * 1000));
}
```
My latest change introduces a better (more mathematically accurate) rounding method than the original solution.

# Test Plan

Before:

1. Load any video into the `<Video />` tag, providing a reference (`videoRef`).
2. Execute the following:

```js
videoRef.current?.setPositionAsync(1001, {
    toleranceMillisBefore: 0,
    toleranceMillisAfter: 0
})
```
3. Observe `positionMillis` inside the `getStatus` method, where we compute the value to be sent to JavaScript.
4. Value is 1000, even though player keeps 1001!

![3](https://github.com/expo/expo/assets/9088288/b8bcdfd7-ad33-4ebf-b65e-9f92c0aa17ab)

After:

1. Load any video into the `<Video />` tag, providing a reference (`videoRef`).
2. Execute the following:

```js
videoRef.current?.setPositionAsync(1001, {
    toleranceMillisBefore: 0,
    toleranceMillisAfter: 0
})
```
3. Observe `positionMillis` inside the `getStatus` method, where we compute the value to be sent to JavaScript.
4. Value is 1001

![precise](https://github.com/expo/expo/assets/9088288/1f53810a-7bd0-4b58-9296-b1fe12e7cad2)

jpudysz commented on packages/expo-av/ios/EXAV/EXAVPlayerData.m:
```diff
@@ -205,6 +205,27 @@ - (BOOL)_shouldPlayerPlay
+    
+  NSNumber *seekMillisToleranceBefore = parameters[EXAVPlayerDataStatusSeekMillisToleranceBeforeKeyPath];
+    
+  if (CMTimeCompare(CMTimeMakeWithSeconds(seekMillisToleranceBefore.floatValue / 1000, NSEC_PER_SEC), kCMTimeZero) != 0) {
```
No, these values are fine. There is no need for precise rounding of tolerances.

tsapeta replied:
I'm also asking about the consistency. I think it makes sense to generally avoid `floatValue` as we already know that it's causing precision issues.

jpudysz replied:
Surely it won't hurt, if you want I can update those values too

tsapeta replied:
Yeah, let's update it then 😉

jpudysz committed:
Merge branch 'expo:main' into main

jpudysz committed:
chore: change floating points from float to double

tsapeta reviewed (approved):
I'm sorry, I totally forgot about this PR 😅 Is this ready to merge?

jpudysz commented:
Yes, it’s ready and working great on my side. There have been no reports from QA for the past 3 weeks 🤞
